### PR TITLE
Example workflow for Loki to enable beamlime

### PR DESCRIPTION
**Pull request by MridulS**
This setups a workflow to address https://github.com/scipp/beamlime/issues/159.

Not too sure where to keep the function but for now I have created a new file.

Fixes scipp/beamlime#159.

### CODE REVIEW

**Comment by MridulS** [comment]
What's the policy about adding dependencies like beamlime here? Should I just add it as a proper dependency or as an optional one as it is only required by a specific code path.

**Comment by SimonHeybrock** [comment]
> What's the policy about adding dependencies like beamlime here? Should I just add it as a proper dependency or as an optional one as it is only required by a specific code path.

We have not actually discussed this specific case. The live-data dashboard will not be run by regular users, but "by us" on a dedicated machine (VM), so personally I would think we can simply *not* add beamlime as a dependency (and install by hand when needed).

File: src/ess/loki/__init__.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 from . import data, general, io
 from .general import default_parameters
+from .workflow import live_workflow
```
Comment (SimonHeybrock) [inline src/ess/loki/__init__.py:8]: Can we avoid the import, such that users can still import this module without having to install Beamlime (see my earlier comment)?

File: src/ess/loki/workflow.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+try:
+    import numpy as np
+    import scipp as sc
+    from beamlime import WorkflowResult
+    from scippneutron.io.nexus.load_nexus import JSONGroup
+except ImportError as e:
+    raise ImportError(
+        "Please install scipp, scippneutron, and beamlime to run workflows with loki."
```
Comment (SimonHeybrock) [inline src/ess/loki/workflow.py:8]: The former two are actually in the dependencies already.

File: src/ess/loki/workflow.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    """Example live workflow function for Loki.
+
+    This function is used to process data with beamlime workflows.
+    """
+    # We do not consume the incoming data right now.
+    _ = group
+    return {
+        'IofQ': sc.DataArray(
+            data=sc.array(dims=['Q'], values=np.zeros(10)),
+            coords={'Q': sc.array(dims=['Q'], values=np.zeros(10), unit='1/angstrom')},
```
Comment (SimonHeybrock) [inline src/ess/loki/workflow.py:22]: I'd make this a `sc.linspace` with reasonable bounds, so plotting will work (for example as used in the some of the examples in the docs: `sc.linspace(dim='Q', start=0.01, stop=0.3, num=101, unit='1/angstrom')`).

**Comment by MridulS** [comment]
> (and install by hand when needed).

Maybe keep this is an optional dependency then? Could be easier to track all deps required for live-data dashboard (if and when more are added) in a single place. Something like `pip install esssans[live]`

File: src/ess/loki/workflow.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+def live_workflow(group: JSONGroup) -> WorkflowResult:
+    """Example live workflow function for Loki.
+
+    This function is used to process data with beamlime workflows.
+    """
+    # We do not consume the incoming data right now.
+    _ = group
+    return {
+        'IofQ': sc.DataArray(
+            data=sc.array(dims=['Q'], values=np.zeros(101)),
```
Comment (SimonHeybrock) [inline src/ess/loki/workflow.py:18]: This should actually be length 100, since Q is typically bin-edges in our applications.

Comment (SimonHeybrock) [reply src/ess/loki/workflow.py:18]: Note there is also `sc.zeros`, which avoids Numpy in this case.

File: src/ess/loki/workflow.py [hunk 5]
```diff
@@ -15,10 +14,10 @@ def live_workflow(group: JSONGroup) -> WorkflowResult:
     _ = group
     return {
         'IofQ': sc.DataArray(
-            data=sc.array(dims=['Q'], values=np.zeros(101)),
+            data=sc.zeros(dims=['Q'], shape=[100]),
             coords={
                 'Q': sc.linspace(
-                    dim='Q', start=0.01, stop=0.3, num=101, unit='1/angstrom'
+                    dim='Q', start=0.01, stop=0.3, num=100, unit='1/angstrom'
```
Comment (SimonHeybrock) [inline src/ess/loki/workflow.py:20]: Sorry, I was unclear: The data has length 100, but the coord length 101, i.e., this is a histogram where `Q` gives the bin edges.

**Review by SimonHeybrock — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
> > (and install by hand when needed).
> 
> Maybe keep this is an optional dependency then? Could be easier to track all deps required for live-data dashboard (if and when more are added) in a single place. Something like `pip install esssans[live]`

You are right, this might be what we do in the long run. But as Beamlime (and in particular this plugin mechanism) is in a very early state we might wait a little.

**Pull request merged**